test/Test/Database/LSMTree/Internal/Vector/Growing.hs

```diff
@@ -0,0 +1,212 @@
+module Test.Database.LSMTree.Internal.Vector.Growing (tests) where
+
+import           Prelude hiding (head, length, tail)
+
+import           Control.Category ((>>>))
+import           Control.Monad.ST.Strict (ST, runST)
+import qualified Data.List as List (length)
+import           Data.Vector (Vector, toList)
+import           Database.LSMTree.Internal.Vector.Growing (GrowingVector,
+                     append, freeze, new)
+import           Test.QuickCheck (Arbitrary (arbitrary, shrink), Gen,
+                     NonNegative (NonNegative, getNonNegative), Positive
+                     (Positive), Property, Small (Small), Testable, chooseInt,
+                     shrinkList, shrinkMap, shrinkMapBy, sized, tabulate, (===))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+-- * Tests
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Vector.Growing" $
+        [
+            testProperty "Final vector is correct"
+                         prop_finalVectorIsCorrect
+        ]
+
+-- * Utilities
+
+-- ** Segments
+
+data Segment a = Replicate Int a deriving stock Show
+
+segmentLength :: Segment a -> Int
+segmentLength (Replicate count _) = max 0 count
+
+segmentToList :: Segment a -> [a]
+segmentToList (Replicate count val) = replicate count val
+
+appendSegment :: GrowingVector s a -> Segment a -> ST s ()
+appendSegment vec (Replicate count val) = append vec count val
+
+-- ** Vector construction
+
+{-
+    Constructs an ordinary vector by creating a growing vector, appending
+    segments to it, and finally freezing it.
+-}
+finalVector :: Int         -- Initial buffer size
+            -> [Segment a] -- Segments
+            -> Vector a    -- Final vector
+finalVector initialBufferSize segments = runST $ do
+    vec <- new initialBufferSize
+    mapM_ (appendSegment vec) segments
+    freeze vec
+
+{-
+    Supplies the final contents of a growing vector for constructing a property.
+
+    The resulting property additionally provides statistics about the buffer
+    size scaling exponents, where a buffer size scaling exponent is the binary
+    logarithm of the ratio between the buffer size after and before appending a
+    segment.
```

Comment:
I think this could be elaborated a bit. What are these exponent numbers supposed to tell me?

Reply:
Hmm, I had hoped that this would be somehow clear in conjunction with the documentation of the `GrowingVector` type. Buffer size scaling exponents tell you if the buffer is grown and, if yes, how much. If you think of growth happening always by doubling the buffer size, a buffer size scaling exponent says how many growth cycles are needed to accommodate the increased vector length. In practice, though, we don’t perform multiple growth cycles for a single append but grow directly to the final buffer size.

What additional explanation would be helpful? Would a reference to the description of the growing approach in the documentation of `GrowingVector` be sufficient perhaps?

Reply:
Just this intuition

> Buffer size scaling exponents tell you if the buffer is grown and, if yes, how much. If you think of growth happening always by doubling the buffer size, a buffer size scaling exponent says how many growth cycles are needed to accommodate the increased vector length.

would be nice to include.